### PR TITLE
blockvolume: Code re-org to fix race condition and cleanup bhv

### PR DESCRIPTION
**Pull request by poornimag**
The block host volume will still exist even when the blocks are all
deleted. Manually deleting block host volume will need another step
of unmounting. With this patch we auto-delete the block host volume
if there are no blocks.

The availablility check of bhv free space is not done within lock, hence
there is a possiblity that the available space has changed by the
time we decide to create the volume. This patch also fixes the race
condition.

Signed-off-by: Poornima G <pgurusid@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
add to whitelist

File: glusterd2/commands/volumes/volume-delete.go [hunk 1]
```diff
@@ -37,39 +39,46 @@ func registerVolDeleteStepFuncs() {
...
+		return
+	}
+
+	events.Broadcast(volume.NewEvent(volume.EventVolumeDeleted, volinfo))
+
+	restutils.SendHTTPResponse(ctx, w, http.StatusNoContent, nil)
+}
+
+// DeleteVolume deletes the volume
+func DeleteVolume(ctx context.Context, volname string) (volInfo *volume.Volinfo, status int, err error) {
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/volume-delete.go]: remove named return as you are returning values

File: glusterd2/commands/volumes/volume-delete.go [hunk 2]
```diff
@@ -37,39 +39,46 @@ func registerVolDeleteStepFuncs() {
 }
 
 func volumeDeleteHandler(w http.ResponseWriter, r *http.Request) {
-
 	ctx := r.Context()
-	logger := gdctx.GetReqLogger(ctx)
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/volume-delete.go]: dont we require this logger?

File: glusterd2/commands/volumes/volume-stop.go [hunk 3]
```diff
@@ -118,35 +119,44 @@ func registerVolStopStepFuncs() {
 func volumeStopHandler(w http.ResponseWriter, r *http.Request) {
 
 	ctx := r.Context()
-	ctx, span := trace.StartSpan(ctx, "/volumeStopHandler")
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/volume-stop.go]: as this is the main function with HTTP handler i think tracer should start from here only

File: glusterd2/commands/volumes/volume-stop.go [hunk 4]
```diff
@@ -118,35 +119,44 @@ func registerVolStopStepFuncs() {
...
+		return
+	}
 
+	events.Broadcast(volume.NewEvent(volume.EventVolumeStopped, volinfo))
+	resp := createVolumeStopResp(volinfo)
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, resp)
+}
+
+// StopVolume will stop the volume
+func StopVolume(ctx context.Context, volname string) (volInfo *volume.Volinfo, status int, err error) {
```
Comment (Madhu-1) [inline glusterd2/commands/volumes/volume-stop.go]: remove named return

File: glusterd2/options/cluster.go [hunk 5]
```diff
@@ -38,6 +38,7 @@ var ClusterOptMap = map[string]*ClusterOption{
 	"auto-create-block-hosting-volumes":  {"auto-create-block-hosting-volumes", "true", OptionTypeBool, nil},
 	"block-hosting-volume-replica-count": {"block-hosting-volume-replica-count", "3", OptionTypeInt, nil},
 	"block-hosting-volume-type":          {"block-hosting-volume-type", "Replicate", OptionTypeStr, nil},
+	"auto-delete-block-hosting-volumes":  {"auto-delete-block-hosting-volumes", "true", OptionTypeBool, nil},
```
Comment (Madhu-1) [inline glusterd2/options/cluster.go]: I think it's better to make default as false or delete BHV if more than one are there, this avoids the frequent creation and deletion of BHV if we keep on creating and deleting single PVC


@aravindavk any thoughts?

File: glusterd2/volume/metadata.go [hunk 6]
```diff
@@ -2,11 +2,13 @@ package volume
 
 const (
 	// BlockHostingAvailableSize is a volume metadata to store available size of hosting volume to create block devices.
-	BlockHostingAvailableSize = "_block-hosting-available-size"
+	BlockHostingAvailableSize = "block-hosting-available-size"
```
Comment (Madhu-1) [inline glusterd2/volume/metadata.go]: is this a metadata field set on BHV volume? if yes metadata should start with `_` or else user can delete metadata

File: glusterd2/volume/metadata.go [hunk 7]
```diff
@@ -2,11 +2,13 @@ package volume
...
 	// BlockHostingVolumeAutoCreated is volume metadata which will be set as `yes` if gd2 auto create the hosting volume
 	BlockHostingVolumeAutoCreated = "block-hosting-volume-auto-created"
 	// BlockHosting is a volume metadata which will be set as `yes' for volumes which are able to host block devices.
 	BlockHosting = "block-hosting"
 	// BlockPrefix is the prefix of the volume metadata which will contain BlockPrefix + blockname as the key and size of the block as value.
 	BlockPrefix = "block-vol:"
+	// BlockHostMarkedForPrune is a metadata that is set to indicate that the bhv is being deleted as it has no block volumes present
+	BlockHostMarkedForPrune = "block-hosting-marked-for-prune"
```
Comment (Madhu-1) [inline glusterd2/volume/metadata.go]: same here also

File: plugins/blockvolume/blockprovider/gluster-block/glusterblock.go [hunk 8]
```diff
@@ -95,45 +90,59 @@ func (g *GlusterBlock) CreateBlockVolume(name string, size uint64, hostVolume st
...
-			hostVol = blockVol.HostVolume()
-			break
-		}
-	}
+}
 
-	if hostVol == "" {
-		return errors.New("block volume not found")
+// DeleteBlockVolume deletes a gluster block volume of give name
+func (g *GlusterBlock) DeleteBlockVolume(name string, options ...blockprovider.BlockVolOption) error {
```
Comment (Madhu-1) [inline plugins/blockvolume/blockprovider/gluster-block/glusterblock.go]: this function and `GetAndDeleteBlockVolume` looks similar, any major difference, can we avoid code duplication?

File: plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go [hunk 9]
```diff
@@ -108,53 +105,53 @@ func (g *GlusterVirtBlk) CreateBlockVolume(name string, size uint64, hostVolume
...
-		return nil, err
+	return &BlockVolume{
+		hostVolume: hostVolume,
+		name:       name,
+		size:       size,
+	}, nil
+}
+
+// GetAndDeleteBlockVolume deletes a gluster block volume of give name
+func (g *GlusterVirtBlk) GetAndDeleteBlockVolume(name string, options ...blockprovider.BlockVolOption) (blockprovider.BlockVolume, error) {
```
Comment (Madhu-1) [inline plugins/blockvolume/blockprovider/gluster-virtblock/gluster_virtblock.go]: same here 

this function and GetAndDeleteBlockVolume looks similar, any major difference, can we avoid code duplication?

File: plugins/blockvolume/hostvol/hostvol_manager.go [hunk 10]
```diff
@@ -143,9 +181,156 @@ func (g *GlusterVolManager) GetOrCreateHostingVolume(name string, minSizeLimit u
...
+
+	if err := volume.AddOrUpdateVolume(volInfo); err != nil {
+		log.WithError(err).Error("failed in updating volume info to store")
+		clusterLocks.UnLock(context.Background())
+		return err
+	}
+	clusterLocks.UnLock(context.Background())
+
+	if prune == true {
+		_ = g.pruneBHV(volInfo.Name, blkName, size)
```
Comment (Madhu-1) [inline plugins/blockvolume/hostvol/hostvol_manager.go]: if we are not logging the error in think we don't need to return an error from pruneBHV

File: plugins/blockvolume/blockprovider/blockvolume_provider.go [hunk 11]
```diff
@@ -20,6 +20,7 @@ var (
 type Provider interface {
 	CreateBlockVolume(name string, size uint64, hostVolume string, options ...BlockVolOption) (BlockVolume, error)
 	DeleteBlockVolume(name string, options ...BlockVolOption) error
+	GetAndDeleteBlockVolume(name string, options ...BlockVolOption) (BlockVolume, error)
```
Comment (oshankkumar) [inline plugins/blockvolume/blockprovider/blockvolume_provider.go]: why we need this?

File: glusterd2/commands/volumes/volume-delete.go (hunk 2 above)
Comment (poornimag) [reply glusterd2/commands/volumes/volume-delete.go]: Not in volumeDeleteHandler, but is now defined in DeleteVolume.

File: plugins/blockvolume/blockprovider/blockvolume_provider.go (hunk 11 above)
Comment (poornimag) [reply plugins/blockvolume/blockprovider/blockvolume_provider.go]: In DeleteVolume() we need the information of the host volume to prune the bhv. DeleteBlockVolume() doesn't return the host vol info, hence the func GetAndDeleteBlockVolume()

File: glusterd2/commands/volumes/volume-stop.go (hunk 3 above)
Comment (poornimag) [reply glusterd2/commands/volumes/volume-stop.go]: What if VoumeStop is called directly? Volume Create and Start had done it this way, hence followed the same

File: glusterd2/options/cluster.go (hunk 5 above)
Comment (poornimag) [reply glusterd2/options/cluster.go]: Alright will set the default to false

**Comment by poornimag** [comment]
@Madhu-1 have addressed most of the comments, is there anything else?

File: plugins/blockvolume/blockprovider/gluster-block/glusterblock.go [hunk 12]
```diff
@@ -11,6 +10,7 @@ import (
 
 	"github.com/gluster/gluster-block-restapi/client"
 	"github.com/gluster/gluster-block-restapi/pkg/api"
+	"github.com/gluster/glusterd2/pkg/errors"
 	log "github.com/sirupsen/logrus"
```
Comment (Madhu-1) [inline plugins/blockvolume/blockprovider/gluster-block/glusterblock.go]: new line

**Comment by Madhu-1** [comment]
@aravindavk PTAL

**Comment by poornimag** [comment]
SelfHeal test failed, not sure why.., is it possible to retrigger the test?

**Comment by Madhu-1** [comment]
retest this please

**Pull request merged**